### ISSUE TITLE
implement trie based wildcard support

### ISSUE THREAD

**Issue by redboltz** [redboltz/mqtt_cpp#686]
See #684 #676 #685 #681 

We have topic_name and topic_filter validatiton functions. They are already been meged.
I already merged tire supporting functions. 

Now, it is a time to implement trie version of wildcard support to test_broker.

@kleunen , could you demonstrate how to use subscription_map and retained_topic map ?

Could you write a code example or pull request ?

I'm not sure sub_con_online and sub_con_offline need to be shared_ptr or not. I remember your previous approach uses shared_ptr for them. Please check it is still required in the current master. 

I guess that you replace 
https://github.com/redboltz/mqtt_cpp/blob/d035c5bca19eaedd72a05c9582bf23fd6ae3d83a/test/test_broker.hpp#L1699-L1700
to subscription_map.

I think that subscription_map needs to have search by the address of endpoint_t and client_id. 
Consider the following scenario.

1. client1 connected to the broker with CleanStart:false SessionExpiryInterval:inifinity
1. client1 subscribed topic1. Broker updates subscription_map.
1. client1 disconnected. Broker need to search subscription_map by the address of endpoint_t and move it to offline one.

It is implemented using multi_index now. See tag_con.

https://github.com/redboltz/mqtt_cpp/blob/d035c5bca19eaedd72a05c9582bf23fd6ae3d83a/test/test_broker.hpp#L1554-L1557

There are some kind of events to be a trigger to access to subscription_map.

1. publish: access to subscription_map by topic_name and get a set of matched sub_con_online or sub_con_offline.
1. subscribe: insert to subscription_map.
1. unsubscribe: remove from subscription_map. The key is topic_filter.
1. disconnect: move or update sub_con_online element in subscription_map. The key is the address of endpoint_t.
1. connect: move or update sub_con_offine element in subscription_map. The key is client_id.
1. timeout: Session expire case. remove sub_con_offline from subscription_map. The key is client_id.

@kleunen , could you tell me how to implement them ?

**Comment by kleunen** [comment]
> @kleunen , could you demonstrate how to use subscription_map and retained_topic map ?

Yes, i have an example broker and a small unit test setup which shows how the maps should be used. How can I send you the code example ?

**Comment by redboltz** [comment]
Please put the code here with cpp decoration as follows:

```cpp
// your code example
int main() {
    subscription_map m;
    ...
}

```

**Comment by kleunen** [comment]
Tsumaranai monodesuga ( hope this is right thing to say :D )

It does say: 
Remaining size should be 1 (root element only)
Remaining size: 4

So something is wrong somewhere, i think this is in the test actually, not the map itself. 

```C++
//
// Created by wkl04 on 10-12-2019.
//
#include "precomp.h"

#include "subscription_map.h"
#include "retained_topic_map.h"

#include <iostream>
#include <chrono>

void TestSingleSubscription()
{
    std::string text = "example/test/A";

    single_subscription_map< std::function< void() > > map;
    map.insert("example/test/A", [] { std::cout << "example/test/A: Subscriber 1" << std::endl; } );

    try {
        map.insert("example/test/A", [] { std::cout << "example/test/A: Subscriber 2" << std::endl; } );
    } catch(std::exception &e)
    {
        // Should fail due to duplicate subscription
        std::cout << e.what() << std::endl;
    }

    auto handle = map.insert("example/test/B", [] { std::cout << "example/test/B" << std::endl; } );
    std::cout << "Handle of new subscription: " << map.handle_to_subscription(handle) << std::endl;
    map.update(handle, [] { std::cout << "example/test/B updated" << std::endl; } );
   // map.remove(handle);

    map.insert("example/+/A", [] { std::cout << "example/+/A: Plus" << std::endl; } );
    map.insert("example/#", [] { std::cout << "example/#" << std::endl; } );

    std::cout << "Test with hash, plus and full path" << std::endl;
    map.find("example/test/A", [](std::function< void() >  const &a) {
        a();
    });

    map.remove("example/+/A");

    std::cout << "Test with hash, full path" << std::endl;
    map.find("example/test/A", [](std::function< void() > const &a) {
        a();
    });

    map.remove("example/#");

    std::cout << "Test with full path only" << std::endl;
    map.find("example/test/A", [](std::function< void() >  const &a) {
        a();
    });

    map.remove("example/test/A");
    // map.remove("example/test/A");
    map.remove("example/test/B");

    std::cout << "No entries exist anymore" << std::endl;
    map.find("example/test/A", [](std::function< void() >  const &a) {
        a();
    });

    std::cout << "Remaining size should be 1 (root element only)" << std::endl;
    std::cout << "Remaining size: " << map.size() << std::endl;
}

void TestMultipleSubscription()
{
    std::string text = "example/test/A";

    multiple_subscription_map<std::string> map;

    try {
        map.insert("doubleslash//#", "Should fail");
    } catch(std::exception &e)
    {
        // Should fail due to empty path usage
        std::cout << e.what() << std::endl;
    }

    map.insert("example/test/A", "example/test/A: Subscriber 1");
    map.insert("example/test/A", "example/test/A: Subscriber 2");
    map.insert("example/test/B", "example/test/B");
    map.insert("example/+/A", "example/+/A: Plus");
    map.insert("example/#", "example/#");
   // map.insert("#", "Root hash");

    std::cout << "Lookup handle of an entry" << std::endl;
    std::cout << "Size: " << map.lookup("example/test/B").size() << std::endl;
    std::cout << "Path: " << map.handle_to_subscription(map.lookup("example/test/B")) << std::endl;

    map.remove(map.lookup("example/test/A"), "example/test/A: Subscriber 1");

    std::cout << "Test with hash, plus and full path" << std::endl;
    map.find("example/test/A", [](std::string const &a) {
        std::cout << a << std::endl;
    });

    /* map.remove("example/+/A", "example/+/A: Plus");

    std::cout << "Test with hash, full path" << std::endl;
    map.find("example/test/A", [](std::string const &a) {
        std::cout << a << std::endl;
    });

    map.remove("example/#", "example/#");

    std::cout << "Test with full path only" << std::endl;
    map.find("example/test/A", [](std::string const &a) {
        std::cout << a << std::endl;
    });

    map.remove("example/test/A", "example/test/A: Subscriber 1");
    map.remove("example/test/A", "example/test/A: Subscriber 2");
    map.remove("example/test/B", "example/test/B");

    std::cout << "No entries exist anymore" << std::endl;
    map.find("example/test/A", [](std::string const &a) {
        std::cout << a << std::endl;
    }); */

    std::cout << "Remaining size should be 1 (root element only)" << std::endl;
    std::cout << "Remaining size: " << map.size() << std::endl;
}

void TestRetainedTopics()
{
    retained_topic_map<std::string> map;
    map.insert_or_update("example/test/A", "example/test/A: Content 1");
    map.insert_or_update("example/test/B", "example/test/B: Content 1");
    map.insert_or_update("example/A/test", "example/A/test: Content 1");
    map.insert_or_update("example/B/test", "example/B/test: Content 1");

    /*map.find("example/test/A", [](std::string const &a) {
        std::cout << a << std::endl;
    });

    map.find("example/test/B", [](std::string const &a) {
        std::cout << a << std::endl;
    });

    map.find("example/test/+", [](std::string const &a) {
        std::cout << a << std::endl;
    });

    map.find("example/+/B", [](std::string const &a) {
        std::cout << a << std::endl;
    }); */

    std::cout << "Match: example/#" << std::endl;
    map.find("example/#", [](std::string const &a) {
        std::cout << a << " " << a.size() << std::endl;
    });
    std::cout << "Done" << std::endl;

    map.remove("example/test/A");
    map.find("example/+/A", [](std::string const &a) {
        std::cout << a << std::endl;
    });

}


int main(int, char**)
{
    try {
        TestSingleSubscription();
        TestMultipleSubscription();
        TestRetainedTopics();


    } catch(std::exception &e)
    {
        std::cout << e.what() << std::endl;
    }
}
``

**Comment by kleunen** [comment]
And the small broker I made: 

```C++
#include "precomp.h"

#include <string>
#include <iostream>

#include "mqtt_server_cpp.hpp"
#include "subscription_map.h"
#include "retained_topic_map.h"

using con_t = MQTT_NS::server<>::endpoint_t;
using con_sp_t = std::shared_ptr<con_t>;

struct session_t;
using session_ptr_t = std::shared_ptr<session_t>;
using session_weak_ptr_t = std::weak_ptr<session_t>;

using session_map_t = std::map<MQTT_NS::buffer, session_ptr_t>;

struct session_subscription_t
{
    session_ptr_t session;
    MQTT_NS::qos qos;

    session_subscription_t(const session_ptr_t &session, MQTT_NS::qos qos = MQTT_NS::qos())
        : session(session), qos(qos)
    { }
};

inline bool operator<(session_subscription_t const &a, session_subscription_t const &b)
{
    return a.session < b.session;
}

using subscription_map_t = multiple_subscription_map< session_subscription_t >;

// Map of all subscription -> session
subscription_map_t subs_map;

struct session_t
{
    MQTT_NS::buffer client_id;
    std::weak_ptr<con_t> con;
    std::map<MQTT_NS::buffer, subscription_map_t::handle> subscriptions;

    session_t(const std::weak_ptr<con_t> &con)
        : con(con)
    { }

    ~session_t()
    {
        std::cout << "Release session: " << client_id << std::endl;
    }

    std::shared_ptr<con_t> get_connection()
    {
        auto sp = con.lock();
        BOOST_ASSERT(sp);
        return sp;
    }

    void add_subscription(MQTT_NS::buffer topic, subscription_map_t::handle const &handle)
    {
        subscriptions[topic] = handle;
    }

    subscription_map_t::handle get_subscription(MQTT_NS::buffer topic)
    {
        auto sub_iter = subscriptions.find(topic);
        if(sub_iter == subscriptions.end())
            return subscription_map_t::handle();
        else
            return sub_iter->second;
    }

    subscription_map_t::handle remove_subscription(MQTT_NS::buffer topic)
    {
        auto sub_iter = subscriptions.find(topic);
        if(sub_iter == subscriptions.end())
            return subscription_map_t::handle();

        auto result = sub_iter->second;
        subscriptions.erase(sub_iter);
        return result;
    }



    void publish(MQTT_NS::buffer topic_name, MQTT_NS::buffer contents, MQTT_NS::publish_options options) {
        auto sp = con.lock();
        if(sp) {
            sp->async_publish(
                    boost::asio::buffer(topic_name),
                    boost::asio::buffer(contents),
                    std::make_pair(topic_name, contents), options);
        }
    }
};


// All stored retained topics
retained_topic_map< std::pair<MQTT_NS::buffer, MQTT_NS::buffer> > retained_map;

// Map of all sessions
session_map_t sessions;

inline void close_session(session_ptr_t session) {
    // Remove all subscriptions for this session
    for(auto const &i: session->subscriptions) {
        subs_map.remove(i.second, session_subscription_t(session));
    }

    // Remove the client id from the list of active sessions
    auto i = sessions.find(session->client_id);
    if(i != sessions.end() && i->second == session)
        sessions.erase(i);

    std::cout << "Active sessions: " << sessions.size() << std::endl;
    std::cout << "Subs map size: " << subs_map.size() << std::endl;
}

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cout << argv[0] << " port" << std::endl;
        return -1;
    }
    boost::asio::io_context ioc;

    auto server_endpoint = boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(),boost::lexical_cast<std::uint16_t>(argv[1]));
    auto s = MQTT_NS::server<>(server_endpoint, ioc);

    s.set_error_handler(
            [](MQTT_NS::error_code ec) {
                std::cout << "error: " << ec.message() << std::endl;
            }
    );


    s.set_accept_handler(
            [](con_sp_t spep) {
                auto& ep = *spep;

                session_ptr_t session = std::make_shared<session_t>(std::weak_ptr<con_t>(spep));

                using packet_id_t = typename std::remove_reference_t<decltype(ep)>::packet_id_t;
                std::cout << "accept" << std::endl;

                // Pass spep to keep lifetime.
                // It makes sure wp.lock() never return nullptr in the handlers below
                // including close_handler and error_handler.
                ep.start_session(session);

                // set connection (lower than MQTT) level handlers
                ep.set_close_handler(
                        [session]() {
                            std::cout << "[server] closed session: " << session->client_id << std::endl;
                            close_session(session);
                        });

                ep.set_error_handler(
                        [session](MQTT_NS::error_code ec) {
                            std::cout << "[server] error: " << ec.message() << " " << session->client_id << std::endl;
                            close_session(session);
                        });

                // set MQTT level handlers
                ep.set_connect_handler(
                        [session](MQTT_NS::buffer client_id, MQTT_NS::optional<MQTT_NS::buffer> username, MQTT_NS::optional<MQTT_NS::buffer> password, MQTT_NS::optional<MQTT_NS::will>, bool clean_session, std::uint16_t keep_alive) {
                            auto sp = session->get_connection();

                            using namespace MQTT_NS::literals;
                            std::cout << "[server] client_id    : " << client_id << std::endl;
                            std::cout << "[server] username     : " << (username ? username.value() : "none"_mb) << std::endl;
                            std::cout << "[server] password     : " << (password ? password.value() : "none"_mb) << std::endl;
                            std::cout << "[server] clean_session: " << std::boolalpha << clean_session << std::endl;
                            std::cout << "[server] keep_alive   : " << keep_alive << std::endl;

                            if(sessions.find(client_id) != sessions.end())
                            {
                                // A session with this client id already exists
                                sp->connack(false, MQTT_NS::connect_return_code::identifier_rejected);
                                return false;
                            }

                            session->client_id = client_id;
                            sessions[client_id] = session;

                            sp->connack(false, MQTT_NS::connect_return_code::accepted);
                            return true;
                        }
                );

                ep.set_pingreq_handler([session]() {
                    auto sp = session->get_connection();
                    sp->async_pingresp();
                    return true;
                });

                ep.set_disconnect_handler(
                        [session]() {
                            auto sp = session->get_connection();
                            std::cout << "[server] disconnect received." << std::endl;
                            close_session(session);
                        });

                ep.set_puback_handler(
                        [](packet_id_t packet_id){
                          //  std::cout << "[server] puback received. packet_id: " << packet_id << std::endl;
                            return true;
                        });

                ep.set_pubrec_handler(
                        [](packet_id_t packet_id){
                          //  std::cout << "[server] pubrec received. packet_id: " << packet_id << std::endl;
                            return true;
                        });

                ep.set_pubrel_handler(
                        [](packet_id_t packet_id){
//                            std::cout << "[server] pubrel received. packet_id: " << packet_id << std::endl;
                            return true;
                        });

                ep.set_pubcomp_handler(
                        [](packet_id_t packet_id){
                           // std::cout << "[server] pubcomp received. packet_id: " << packet_id << std::endl;
                            return true;
                        });

                ep.set_publish_handler(
                        [session]
                                (MQTT_NS::optional<packet_id_t> packet_id,
                                 MQTT_NS::publish_options pubopts,
                                 MQTT_NS::buffer topic_name,
                                 MQTT_NS::buffer contents){
                            /* std::cout << "[server] publish received."
                                      << " dup: "    << pubopts.get_dup()
                                      << " qos: "    << pubopts.get_qos()
                                      << " retain: " << pubopts.get_retain() << std::endl;
                            if (packet_id)
                                std::cout << "[server] packet_id: " << *packet_id << std::endl;
                            std::cout << "[server] topic_name: " << topic_name << std::endl;
                            std::cout << "[server] contents: " << contents << std::endl; */

                            if(!mqtt_valid_topic(topic_name))
                                return false;

                            if(pubopts.get_retain() == MQTT_NS::retain::yes)
                                retained_map.insert_or_update(topic_name, std::make_pair(topic_name, contents));

                            std::map< session_ptr_t, MQTT_NS::qos > subscribers;
                            subs_map.find(std::string(topic_name.data(), topic_name.size()), [&topic_name, &contents, &pubopts, &subscribers](session_subscription_t const &r){
                                subscribers[r.session] = std::max(r.qos, subscribers[r.session]);
                            });

                            for(auto const &r: subscribers)
                                r.first->publish(topic_name, contents, std::min(r.second, pubopts.get_qos()) |  pubopts.get_retain());

                            return true;
                        });

                ep.set_subscribe_handler(
                        [session] (packet_id_t packet_id, std::vector<std::tuple<MQTT_NS::buffer, MQTT_NS::subscribe_options>> entries) {
                            std::cout << "[server]subscribe received. packet_id: " << packet_id << std::endl;
                            std::vector<MQTT_NS::suback_return_code> res;
                            res.reserve(entries.size());

                            auto sp = session->get_connection();

                            for (auto const& e : entries) {
                                MQTT_NS::buffer topic = std::get<0>(e);
                                MQTT_NS::qos qos_value = std::get<1>(e).get_qos();

                                if(mqtt_valid_subscription(topic)) {

                                    auto handle = session->get_subscription(topic);
                                    if(!handle.empty())
                                        subs_map.update(handle, session_subscription_t(session, qos_value));
                                    else
                                        session->add_subscription(topic, subs_map.insert(topic, session_subscription_t(session, qos_value)));

                                    res.emplace_back(MQTT_NS::qos_to_suback_return_code(qos_value));

                                    retained_map.find(topic, [&session, &topic, qos_value](std::pair<MQTT_NS::buffer, MQTT_NS::buffer> const &retained_message) {
                                        session->publish(retained_message.first, retained_message.second, qos_value);
                                    });
                                } else {
                                    res.emplace_back(MQTT_NS::suback_return_code::failure);
                                }


                            }

                            sp->suback(packet_id, res);
                            return true;
                        }
                );

                ep.set_unsubscribe_handler([session](packet_id_t packet_id, std::vector<MQTT_NS::buffer> topics) {
                            auto sp = session->get_connection();
                            std::cout << "[server]unsubscribe received. packet_id: " << packet_id << ", client id: " << session->client_id << std::endl;

                            for (auto const& topic : topics) {
                                auto handle = session->remove_subscription(topic);

                                subs_map.remove(handle, session_subscription_t(session));
                            }

                            sp->unsuback(packet_id);
                            return true;
                        }
                );
            }
    );

    s.listen();

    ioc.run();
}

**Comment by kleunen** [comment]
If you really consider serializing the retained and published messages to disk, i would recommend using an embedded database such as: [LevelDB](https://github.com/google/leveldb) or[KyotoCabinet](https://en.wikipedia.org/wiki/Tokyo_Cabinet_and_Kyoto_Cabinet). And perform the actual writing on different thread then the network/mqtt handling.

**Comment by redboltz** [comment]
Thanks you! I will take a look your codes.

> If you really consider serializing the retained and published messages to disk, i would recommend using an embedded database such as: [LevelDB](https://github.com/google/leveldb) or[KyotoCabinet](https://en.wikipedia.org/wiki/Tokyo_Cabinet_and_Kyoto_Cabinet). And perform the actual writing on different thread then the network/mqtt handling.

Yes, I wrote examples on wiki https://github.com/redboltz/mqtt_cpp/wiki/Persistence
And my proprietary broker (not mqtt_cpp one) uses redis to store message.

> Not sure how easy it would be to add this to the design now, that is something to consider for the future ?

test_broker.hpp is originally designed for testing for client code.
Now, it is becoming one of reference implementation of MQTT v5 broker.
Perhaps test_broker.hpp is moved to something like `utility/simple_broker.hpp` .

**Comment by redboltz** [comment]
> Tsumaranai monodesuga ( hope this is right thing to say :D )
It is perfect! つまらないものですが 

Thank you for interested in Japanese :)

**Comment by redboltz** [comment]
https://github.com/redboltz/mqtt_cpp/blob/a897c81932a0b363388b1bc53a56b8ab61175538/test/subscription_map.hpp#L390-L393

It seems that the function insert() calls insert_or_update(). Is this right?
I think that insert() should be failed if the same element exists.
The return value is similar to https://en.cppreference.com/w/cpp/container/set/insert
maybe `std::pair<handle, bool>`.

**Comment by redboltz** [comment]
I need to know newly inserted or updated.

If subscription option retain_handling is send_only_new_subscription, then send retained message to the subscriber only if the subscription is new one.

See

https://github.com/redboltz/mqtt_cpp/blob/a897c81932a0b363388b1bc53a56b8ab61175538/test/test_broker.hpp#L981-L1012

https://github.com/redboltz/mqtt_cpp/blob/a897c81932a0b363388b1bc53a56b8ab61175538/test/test_broker.hpp#L1073-L1098

**Comment by kleunen** [comment]
> maybe `std::pair<handle, bool>`

I agree, but an optional is better in this case. Also, the insert in the single subscription map throws an exception if a duplicate store is detected. I will also change that to returning an optional.

**Comment by kleunen** [comment]
> https://github.com/redboltz/mqtt_cpp/blob/a897c81932a0b363388b1bc53a56b8ab61175538/test/subscription_map.hpp#L390-L393
> 
> It seems that the function insert() calls insert_or_update(). Is this right?
> I think that insert() should be failed if the same element exists.
> The return value is similar to https://en.cppreference.com/w/cpp/container/set/insert
> maybe `std::pair<handle, bool>`.

I don't think it should fail when the element already exists. The c++ container set also does not really fail, it just states whether it inserted the element or it is returning the existing element. So actually the set insert is insert_or_update. In case of a single subscription map, it should fail.

**Comment by redboltz** [comment]
Ok, I come up with a new question. What is the difference between single_subscription_map and multiple_subscription_map?
I think that it is similar to map and multi_map relationship.

I think that multiple_subscription_map::insert_or_update() is quite misleading.Because it always insert never update. Is that right?
I think that multiple_subscription_map::insert_or_update() should be removed.
How about multiple_subscription_map::update() ? I'm not sure the behavior.

single_subscription_map::insert_or_update()  doesn't exist. I think that it should exist.

What do you think?

**Comment by kleunen** [comment]
> I think that multiple_subscription_map::insert_or_update() is quite misleading.Because it always insert never update. Is that right?

yes, exactly, insert_or_update should be renamed to insert, same as in a set.

**Comment by kleunen** [comment]
> How about multiple_subscription_map::update() ? I'm not sure the behavior.

The update with the handle, inserts the value if the subscription already exists. 
Not sure what to name this, or maybe just remove.

**Comment by redboltz** [comment]
Thanks.

In multiple_subscription_map, I think that the following update is good:

1. remove `handle insert(MQTT_NS::string_view subscription, Value const& value)`.
1. rename `handle insert_or_update(MQTT_NS::string_view subscription, Value value)` to `handle insert(MQTT_NS::string_view subscription, Value value)`
1. rename `void update(handle h, Value value)` to void insert(handle h, Value value)`.

If it is OK, I will update. And then I start implementing test_broker using #687  way.

How about single_subscription_map ?

BTW, I only use multiple_subscription_map in test_broker.hpp. So single_subscription_map is not so important for me.

**Comment by kleunen** [comment]
@redboltz Please have a look at #688

**Comment by kleunen** [comment]
The remove returns a boost::optional<Value>, the value that is returned is actually the value that you give when calling the function, and the optional indicates whether a value was actually removed or not. If i look at the std::set, the operation is called erase and not remove, and actually the result is a count of the number of items removed:

_(2) size_type erase (const value_type& val);
For the value-based version (2), the function returns the number of elements erased.
Member type size_type is an unsigned integral type._

This actually makes more sense

**Comment by kleunen** [comment]
Now I added the remove -> erase rename to #688, not sure if that was a good idea.

**Comment by redboltz** [comment]
I will ask you rebase #688 from #687 later. I'm fixing #687,

**Comment by redboltz** [comment]
https://github.com/redboltz/mqtt_cpp/pull/687/checks?check_run_id=1263724124
I stuck the error. It is caused by hash related issue required by subscription_map requirement .
Any ideas?

**Comment by kleunen** [comment]
> Any ideas?

hash_value is not implemented for boost::basic_string_view<char, std::char_traits<char>, so I guess we need to give implementation of this function for hashing

**Comment by kleunen** [comment]
Or include boost_string_view header?

**Comment by redboltz** [comment]
It seems that boost::basic_string_view has it
https://www.boost.org/doc/libs/1_74_0/boost/utility/string_view.hpp

**Comment by kleunen** [comment]
Yes, I noticed. Maybe include it explicitely ?

**Comment by redboltz** [comment]
I found the reason.
hash_value function has been introduces since 1.69.0.
Maybe I can fix the problem.

**Comment by kleunen** [comment]
Give implementation of hash_value function, boost version < 1.69

**Comment by kleunen** [comment]
The unit test should include the checking subscription is removed when it is empty in multiple_subscription_map, when all the values are removed from a subscription, the subscription can be removed completely from the map. Also the reference counting should be checked in the unit test. In my test I included checking that when all subscriptions are removed, you only have the root element left: all entries are removed from the hash map, expect the root element

**Comment by kleunen** [comment]
I think the reason was why i used a shared_ptr, is that the connections are not only indexed by the multiple_subscription_map, but by other keys in a multi index as well. You can not add the multiple_subscription_map to another multi_index structure, so I used a shared_ptr. In case a connection disconnects, i remove all the subscriptions from the multiple_subscription_map and this removes all shared_ptr's to the connection. Finally, the last shared_ptr is removed from the multi_index.

**Comment by kleunen** [comment]
In my test broker, multiple subscriptions may point to the same session. Therefore I use a shared_ptr. The session itself (shared_ptr to session), is also attached to the lambda functions for handling incoming messages for a connection.

**Comment by redboltz** [comment]
Thank you. In test_broker.hpp design, multiple con_sp_t are shared by sub_con_online, so sub_con_sp_t is a shared_ptr.
Each subscription has one sub_con_online element. So it doesn't need to be a shared_ptr.
One sub_con_online has one entry in subscription_map. So I choose reference_wrapper<sub_con_online> for subscription_map.

**Comment by redboltz** [comment]
I merged #687. I renamed `subscription_map.cpp` to `subscription_map_broker.cpp`.
If you have a time, please add your tests for `subscription_map` as `subscription_map.cpp`.

**Comment by kleunen** [comment]
I understand, but I would never use reference_wrapper. The pointer in reference_wrapper may become invalid. Cleanup of the subscription_map should be done before the referenced pointer is freed. This complicates cleaning up the data, in case of exceptions, or possibly when code is modified later an error is made. Usage of weak_ptr, reference_wrapper, raw pointers, etc... in almost all cases indicate a mistake in ownership of data. 

Ownership of sub_con_online is shared between entries in subscription_map and mi_sub_con_online, therefore a shared_ptr makes sense. Removing entries from subscription_map and mi_sub_con_online is now independent on order of removal. 

A language such as Rust would not allow such a construct, lifetime of both pointers may only partially overlap and one pointer may outlive lifetime of referenced object. 

Hopefully you understand and appreciate my explanation.

**Comment by redboltz** [comment]
I understand your design policy. But I don't choose it.
I know how dangling reference is dangerous. 

In this case, it is not so difficult to ensure the lifetime.
Maybe sub_con_online destructor, if handle is not empty (perhaps this check is not required if adding subscription_map is done in the constructor parameter initializer), then erase the handle from subscription_map.

Anyway, I will implement broker by my way.

**Comment by redboltz** [comment]
@kleunen , I found a problem.

When topic filter is overlapped, I got "Invalid handle was specified" exception on erase.
Here is a minimal code that reproduces the error.

"overlapped" means subscription topic filter are "a/b/c" and "a/b" relationship. They should be independent. 
But in the directory metaphor, "a/b" is the parent directory of "a/b/c".

What do you think? 

```cpp
    using func_t = std::function<void()>;
    using value_t = std::shared_ptr<func_t>; // shared_ptr for '<' and hash
    using sm_t = multiple_subscription_map<value_t>;

    sm_t m;
    auto v1 = std::make_shared<func_t>([] { std::cout << "v1" << std::endl; });
    auto v2 = std::make_shared<func_t>([] { std::cout << "v2" << std::endl; });

    auto it_success1 = m.insert("a/b/c"_mb, v1);
    assert(it_success1.second);
    auto it_success2 = m.insert("a/b"_mb, v2);
    assert(it_success2.second);

    auto e1 = m.erase(it_success1.first, v1);
    std::cout << e1 << std::endl;
    auto e2 = m.erase(it_success2.first, v2); //  Invalid handle was specified is thrown here
    std::cout << e2 << std::endl;
```

**Comment by redboltz** [comment]
I got m.dump().

### after v1 is inserted
```
2 c 1 1
1 b 0 1
0 a 0 1
18446744073709551615  0 1
```

### after v2 is inserted
```
2 c 1 1
1 b 1 1
0 a 0 1
18446744073709551615  0 1
```

### after v1 is erased
```
18446744073709551615  0 1
```

**Comment by kleunen** [comment]
I will have a look, that is not correct indeed

**Comment by redboltz** [comment]
Thanks!
It happens not only subscription_map but also retained_topic_map.

**Comment by kleunen** [comment]
I am a bit lost now what is the branch / PR i should work on ?

**Comment by redboltz** [comment]
Now, subscription_map and retaind_topic_map are merged into the master.
If you fix the issue, then please create a new PR against master.

The problem is happened on my local environment. Not yet pushed.

**Comment by kleunen** [comment]
But the _mb is to make the string multibyte ? I added a test for the subscription_map, but the _mb is not part of the main project ?

**Comment by kleunen** [comment]
ah yes, we broke the insert.

**Comment by redboltz** [comment]
`_mb` is postfix of `mqtt::buffer`.

**Comment by redboltz** [comment]
> But the _mb is to make the string multibyte ? I added a test for the subscription_map, but the _mb is not part of the main project ?

Now, broker oriented subscription_map test is in subscription_map_broker.cpp.
There is no pure subscription_map tests in the master. If you add it, it is very welcome. Please use the filename subscription_map.cpp for that.

**Comment by kleunen** [comment]
Yes, I added the test, with your broken case. #558 
and the fix

**Comment by redboltz** [comment]
Sorry ? I'm not sure what's going on.
I have no plat to merge #558 
The master has already have subscription_map and retainded_topic_map.
Please add the fix against the master and add tests.

**Comment by kleunen** [comment]
Ok. I made mistake somewhere while pushing to github. Allow me to make proper PR

**Comment by kleunen** [comment]
What about naming in retained_topic_map ? Should we keep insert_or_update, or make this insert also ? 
It is actually an insert_or_update, but makes sense to use the same names

**Comment by redboltz** [comment]
The current interface is

```cpp
template<typename Value>
class retained_topic_map {

    retained_topic_map();

    // Insert a value at the specified subscription path
    void insert_or_update(MQTT_NS::string_view topic, Value const& value);

    // Find all stored topics that math the specified subscription
    void find(MQTT_NS::string_view subscription, std::function< void (Value const&) > const& callback) const;

    // Remove a stored value at the specified topic
    void remove(MQTT_NS::string_view topic);
;
```

I think that the following is good.

```cpp
template<typename Value>
class retained_topic_map {

    retained_topic_map();

    // return inserted or updated object iterator (I'm not sure the type. Something like iterator type)
    // value as forwarding reference. User can move (or pass a temporary value) into the map 
    template <typename V>
    iterator insert_or_update(MQTT_NS::string_view topic_name, V&& value);

    // Find all stored topics that math the specified subscription
    void find(MQTT_NS::string_view topic_filter, std::function< void (Value const&) > const& callback) const;

    // Remove a stored value at the specified topic
    // return value is 0 or 1. But std::size_t is good. 
    // It is out of MQTT spec but if we would have multple_retained_topic_map, then std::size_t is consistent type
    std::size_t erase(MQTT_NS::string_view topic_name);
;
```

**Comment by kleunen** [comment]
I added the unit test and fixed a number of bugs in the reference counting. Updating the interface i would like to do later. Have a look at PR #689

**Comment by redboltz** [comment]
Thank you! I just added review comments.

**Comment by kleunen** [comment]
> ```c++
>     // return inserted or updated object iterator (I'm not sure the type. Something like iterator type)
>     // value as forwarding reference. User can move (or pass a temporary value) into the map 
>     template <typename V>
>     iterator insert_or_update(MQTT_NS::string_view topic_name, V&& value);
> ```

Yes, the retained_topic_map does not have the handles to topics like the subscription_map does. You can return a handle here, which is smaller to store than the complete topic_name.

**Comment by kleunen** [comment]
I am not sure if it is really worthwile for retained_topic_map. For the subscription_map, you have to clear the subscriptions quite often, when the client disconnects. But the retained_topics are always retained, you do not have to store references to these topics in the broker.

**Comment by redboltz** [comment]
Returning iterator by insert_or_update() is not important for test_broker implementation.
I have no problem with the current interface.

If it is easy and cheap operation, it's nice to have. But it isn't, just skip it.

**Comment by kleunen** [comment]
The forwarding references are useful, but should we add that also to the erase operation ? This is now a Value const &, but an rvalue can not be passed to this function now.

**Comment by redboltz** [comment]
If you add (or modify) forwarding reference to insert_or_update() it is nice.
It can reduce copy operation.

For erase, no value parameter is required.
I'm not sure what is the meaning of value parameter because **single** retained_topic_map has at most one value for one key (topic_name).

**Comment by kleunen** [comment]
> I'm not sure what is the meaning of value parameter because **single** retained_topic_map has at most one value for one key (topic_name).

single_value_map has no Value parameter in erase. Returned value can be either 1 or 0, a boolean could also have been used. But I guess the std::set also uses a size_t on erase:

```C++
class single_subscription_map {
    std::size_t erase(MQTT_NS::string_view subscription);
}

**Comment by kleunen** [comment]
> I'm not sure what is the meaning of value parameter because **single** retained_topic_map has at most one value for one key (topic_name).

erase in retained_topic_map does not have Value parameter, same for single_subscription_map. Return value (for both erase operations) is a size_t which is either 1 or 0. This is the same as for multiple_subscription_map and std::set. So this is OK, i would say

```C++
class retained_topic_map {
    std::size_t erase(MQTT_NS::string_view topic);
}

class single_subscription_map{
    std::size_t erase(MQTT_NS::string_view subscription);
    std::size_t erase(handle h);
}

**Comment by redboltz** [comment]
I'm thinking that I use `retained_topic_map` for manage actual data `struct retain`, not view. It is very different from subscription_map usercase.

Subscribe related operations require multiple keys if I want to do the operations efficiently. As I commented before, when a connection that is SessionExpiryInterval:0 is disconnected, all subscription relates to the connection need to be unsubscribed.
It requires searching container by connection. subscription_map doesn't provide it. So I use multi_index for that and subsctiption_map is only for wildcard matching as view class.

But retained_topic_map is different. The operation is add/remove on publish received. If the contents is empty, remove, otherwise add. Find by topic filter happens when subscribe is received. So the all operation is covered by retained_topic_map efficiently.

**Comment by kleunen** [comment]
And MQTT_NS::string_view subscription should be passed by copy right ? 

So: MQTT_NS::string_view subscription
and not: MQTT_NS::string_view subscription const & ?

**Comment by redboltz** [comment]
I think that 

> And MQTT_NS::string_view subscription should be passed by copy right ?

Yes.

**Comment by kleunen** [comment]
> Subscribe related operations require multiple keys if I want to do the operations efficiently. As I commented before, when a connection that is SessionExpiryInterval:0 is disconnected, all subscription relates to the connection need to be unsubscribed.
> It requires searching container by connection. subscription_map doesn't provide it. So I use multi_index for that and subsctiption_map is only for wildcard matching as view class.

That is what the handle was envisioned for: fast removal of subscriptions. So for every Session you can store a list of all subscription handles that are in the subscription map. And when Session is removed, remove all the subscriptions by handle.

**Comment by redboltz** [comment]
I got confused.

I think that as follows:

```cpp
using sm_t = multiple_subscription_map<int>;
sm_t m;
int v1 = 1;
int v2 = 2;
auto [h1, r1] = m.insert("a/b", v1);
auto [h2, r2] = m.insert("a/b", v2);

// I think that the following two function causes the same result but 
// handle version is more efficient (skip searching process)
m.erase(h1, v1);
m.erase("a/b", v1);

```

Is that right?

**Comment by kleunen** [comment]
Yes, 

```C++
m.erase(h1, v1); // Is more efficient
m.erase("a/b", v1); // Is less efficient
```

But also storing h1 is more efficient to store, than storing "a/b", memory size is smaller.

**Comment by redboltz** [comment]
And one session can store multiple handles that is corresponding to subscription.
Then I can call as follows:

```cpp
for (auto const& h : session.handles) {
    m.erase(h, ?);
}
```

And it is efficient. Is that right?

**Comment by kleunen** [comment]
> And it is efficient. Is that right?

Yes, exactly. And ? is a reference to the session itself.

**Comment by redboltz** [comment]
I use handle for that purpose in test_broker.hpp to manage subscription.

But retained_topic_map, I don't use the handle version of erase(). (It is not provided though)

For implement test_broker retain mechanism, `erase("a/b");` is good enough.
Because there is no other related containers.

**Comment by redboltz** [comment]
So I requested adding handle return value for `retained_topic_map::insert_or_update()`, but I don't need it actually. Just for consistent interface. So it is low priority.

**Comment by kleunen** [comment]
CI breaks now :) with perfect-forwarding

**Comment by redboltz** [comment]
https://github.com/kleunen/mqtt_cpp/blob/sub_test_3/test/subscription_map.hpp#L439
V&& is forwarding reference not rvalue reference. So you need to use `std::forward<V>(value)` instead of `std::move(value)`.

**Comment by kleunen** [comment]
> It requires searching container by connection

Yes, you can not search through the map based on value. Also, when storing the handle of a subscription, the complete path through the tree is stored. But actually only the leaf is required. You can trace back the parent. So some performance can still be gained for this operation.

**Comment by kleunen** [comment]
Please consider the performance optimizations in #690, it removes a lot of temporaries when performing certain operations and makes the handle smaller (only the key of the leaf in the subscription tree).

**Comment by kleunen** [comment]
Would be nice to have some public server for this broker as well, once the trie based wildcard support is in.

**Comment by kleunen** [comment]
Regarding the retained topic, there is still an optimization that can be made. A map is now used, the entries are ordered. This is because for a wildcard (#/+), we need to find all topics that match under a certain child (a/#, search all childs of a). But in case we need to lookup a specific child (find a/b), we could use a hash map lookup. This would be possible if we use a multiindex for the retained topics. Use a hash for named lookups (like in the subcription map), and use a ordered_non_unique based on the id of the parent root for evaluating wildcards.

**Comment by redboltz** [comment]
> Would be nice to have some public server for this broker as well, once the trie based wildcard support is in. Like the mosquitto test server: [Mosquitto Test](https://test.mosquitto.org/)

Yes, it is nice to have. But I don't know where is appropriate place.
I personally don't have a plan using my personal AWS (EC2 or container).

And now, MQTTv5 is not completely implemented. I'm implementing it. When all features are implemented, it is a time to publish.

**Comment by redboltz** [comment]
> Regarding the retained topic, there is still an optimization that can be made. A map is now used, the entries are ordered. This is because for a wildcard (#/+), we need to find all topics that match under a certain child (a/#, search all childs of a). But in case we need to lookup a specific child (find a/b), we could use a hash map lookup. This would be possible if we use a multiindex for the retained topics. Use a hash for named lookups (like in the subcription map), and use a ordered_non_unique based on the id of the parent root for evaluating wildcards.

Additional optimizations are very welcome!
I'm integrating trie based map into the broker. If the interface is not changed, it is better but I can accept some changes.

**Comment by redboltz** [comment]
mqtt_cpp supports C++14 or later. When you use boost::multi_index, please don't use https://www.boost.org/doc/libs/1_74_0/libs/multi_index/doc/reference/key_extraction.html#key

It requires C++17.

I sometimes asked why you support C++14 ?
My policy is support one older version of the current up to date version.
Now, C++17 is up to date version. I know about C++20 but not all major compilers support it yet.
If most of major compilers support C++20, then I would update mqtt_cpp minimum standard requirement to C++17.

**Comment by kleunen** [comment]
I will fix size first i think. That way you can add some debugging to the broker. Maybe publish to $sys

**Comment by kleunen** [comment]
I implemented the calculation of the number of subscriptions / retained topics in #695. Also, the PR contains another fix for a reference counting issue in the retained topic map. Additional tests were added

**Comment by kleunen** [comment]
> Yes, it is nice to have. But I don't know where is appropriate place.
> I personally don't have a plan using my personal AWS (EC2 or container).

I was thinking maybe I get a [Pine64 Rock64](https://www.pine64.org/devices/single-board-computers/rock64/) and host from home. I have 1Gbit glassfiber connection.

**Comment by jonesmz** [comment]
> > Would be nice to have some public server for this broker as well, once the trie based wildcard support is in. Like the mosquitto test server: [Mosquitto Test](https://test.mosquitto.org/)
> 
> Yes, it is nice to have. But I don't know where is appropriate place.
> I personally don't have a plan using my personal AWS (EC2 or container).
> 
> And now, MQTTv5 is not completely implemented. I'm implementing it. When all features are implemented, it is a time to publish.

Options: AWS free tier -- I do not know how this works but its worth considering.

Host on contributor connection -- risky if the connection is consumer grade, but viable if business class.

Create a digital ocean market place virtual machine template -- allows people to have a "one click" installation to try out themselves.

Partner with company that uses mqtt_cpp. -- if I were still at my previous job I would offer to do this, so there may be companies that use mqtt_cpp that would be willing to do this.

----------

Concerns:

Need to implement bandwidth limiting and quotas to avoid abuse. This would include persisting statistic information.

Need to implement some kind of security policy modules if hosted by third party. E.g. support policies to allow limiting who can subscribe to which topics based on clientid / auth negotiation / group membership specified by some database or config file. This is important for a company so that they can feel like they can secure the broker against malicious use. Even in a "its just a demonstration server" situation, no company wants to host a service that has zero security controls.

Should convert broker from sync-only to async-only. Important for performance when you have >hundreds of clients.

Test_broker currently does not try to support multiple threads -- if it were being hosted for demonstration purposes, it should use #threads equal to the hardware its running on to demonstrate the high throughput its capable of.

Consider adding some hooks for operating the test broker in a cluster environment, like kubernates. This includes supporting health checking functions like "$healthcheck/kubernates" or whatever topic. It doesn't mean that you'd need to operating more than one instance of the test broker at a time, but just supporting the health checking logic that kubernates needs to understand if it needs to restart the application. Or migrate it to another higher capacity node.

**Comment by kleunen** [comment]
> Concerns:
> 
> Need to implement bandwidth limiting and quotas to avoid abuse. This would include persisting statistic information.
> 

This can be handled at OS / Router level, you don't have to build this into the broker. 

> Need to implement some kind of security policy modules if hosted by third party. E.g. support policies to allow limiting who can subscribe to which topics based on clientid / auth negotiation / group membership specified by some database or config file. This is important for a company so that they can feel like they can secure the broker against malicious use. Even in a "its just a demonstration server" situation, no company wants to host a service that has zero security controls.
> 

I use mosquitto now in production environment, indeed authorization and security is the biggest thing missing now. 

> Should convert broker from sync-only to async-only. Important for performance when you have >hundreds of clients.
> 

Definitly required to have good performance. 

> Test_broker currently does not try to support multiple threads -- if it were being hosted for demonstration purposes, it should use #threads equal to the hardware its running on to demonstrate the high throughput its capable of.
>

This you don't have to do in the broker itself. When you can do is to use the SO_REUSE to provide load balancing over multiple broker processes. See:[SO_REUSE]( https://tech.flipkart.com/linux-tcp-so-reuseport-usage-and-implementation-6bfbf642885a)
 
For example you can start 4 broker processes: broker[0], broker[1], broker[2], broker[3]
Now, with the SO_REUSE, the clients connect to the 4 different brokers. They subscribe to there respective tokens there. 

Only thing is that when 1 client publishes a message, it will be received by 1 broker only, for example broker[0]. What you need is to have 1 republisher client, which connects to all 4 brokers and subscribes to all messages published. And this client republishes every message received on a broker on all other brokers. 

You need to have 1 port which connects to all brokers in the cluster:
port 10000 -> broker[0], broker[1], broker[2], broker[3]

And other ports which connect to specific brokers:
port 10001 -> broker[0]
port 10002 -> broker[1]
port 10003 -> broker[2]
port 10004 -> broker[3]

For the republishing client to use. Not sure how to extend this to shared subscriptions, I guess shared subscriptions need to be republished in a round-robin fashion until one broker has handled the message.  

> Consider adding some hooks for operating the test broker in a cluster environment, like kubernates. This includes supporting health checking functions like "$healthcheck/kubernates" or whatever topic. It doesn't mean that you'd need to operating more than one instance of the test broker at a time, but just supporting the health checking logic that kubernates needs to understand if it needs to restart the application. Or migrate it to another higher capacity node.

Definitely useful

**Comment by kleunen** [comment]
The PR with updated indices compiles and tests OK now on my CI, i can open new PR. Issue was the boost version check in the header was not OK

**Comment by kleunen** [comment]
Regarding the hosting I am seriously considering upgrading my NAS and maybe providing the ability to host a docker image from my home. I have a readynas duo which is running for 10 years now, and needs upgrading anyway. I think the github build system can build a docker image, right ? This should be easy to host. 

Regarding the authentication, I was thinking maybe the broker can provide authentication plugins by publishing authentication requests to a specialized topic. A client can then authenticate the request and give the broker a authenticated message back to the broker. Both the broker and the authentication client will have to use a shared secret key, to make sure the authentication client is authenticated to authorize other clients. 

I think this is a pretty convenient approach. You can write an authentication client in PHP for example and retrieve the authentication from a database or a file. The broker still needs some basic authentication to allow for this client to access the broker. 

Is this something interesting to work out ?

**Comment by kleunen** [comment]
Bug #128 can be closed now right ?

**Comment by redboltz** [comment]
> Bug #128 can be closed now right ?

Ah, yes I just did it.
